### ISSUE TITLE
3 Renaissance benchmarks fail to run with OpenJ9

### ISSUE THREAD

**Issue by smlambert** [eclipse/openj9#6168]
We are looking at adding some Renaissance perf benchmarks, an initial run of ~19 of them (with no special tuning or JVM options) shows that 3 fail to run.  The PR that contains those targets has not yet been merged, the sandbox Grinder is pulling from PR branch at the moment (can rerun the Grinder_SandboxExtra job linked below, or setup internal Grinder to reproduce, if required).

Details at: https://ci.adoptopenjdk.net/view/work%20in%20progress/job/Grinder_SandboxExtra/4/tapResults/

Dotty fails to compile with AssertionError:  
[dottyFailureOutput.txt](https://github.com/eclipse/openj9/files/3301567/dottyFailureOutput.txt)

page-rank - crashes OutOfMemoryError, NullPointerException:
[page-rankFailureOutput.txt](https://github.com/eclipse/openj9/files/3301672/page-rankFailureOutput.txt)

reactors - crashes OutOfMemoryError:
[reactorsFailureOutput.txt](https://github.com/eclipse/openj9/files/3301669/reactorsFailureOutput.txt)

Core files can be found in:
https://ci.adoptopenjdk.net/view/work%20in%20progress/job/Grinder_SandboxExtra/4/artifact/perf_test_output.tar.gz

**Comment by fjeremic** [comment]
@NigelYiboYu can you comment on your experience with this? Did you encounter similar failures when you ran on Z?

**Comment by charliegracie** [comment]
I am trying to reproduce the failures on my OS X machine using an JDK + OpenJ9.
1. dotty runs to completion but I get lots of these messages:
`failure to convert Constant(false) to TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Byte)`
2. page-rank worked for me 5 out of 5 times.  The noticable difference to me in the output was in my run I got the following: `19/06/18 10:20:17 INFO MemoryStore: MemoryStore started with capacity 2.2 GB` and in the test run linked above I see: `19/06/17 19:16:54 INFO MemoryStore: MemoryStore started with capacity 127.2 MB`.  I did not add any command line options. I wonder if the machine it ran on had a very small amount of physical memory?
3. reactor just worked as well.. I guess my question from 2 is my suspicion for this one as well.

**Comment by charliegracie** [comment]
2. I re-ran with different `-Xmx` and `-Xms` options and the default MemoryStore size seems to be about 1/4-1/2 of the heap size. If I run with `-Xms512m` and `-Xmx512m` I get the same 127.2MB as in the failing run and I also get an OOM.

**Comment by charliegracie** [comment]
3. I also get an OOM on reactors if I run with `-Xmx512m` and `-Xms512m`

**Comment by smlambert** [comment]
Thanks, I will suspect/assume insufficient machine resource and disable those 2 for the moment (page-rank and reactor).

**Comment by smlambert** [comment]
I will disable all 3 of them for now under: https://github.com/AdoptOpenJDK/openjdk-tests/issues/1184, so that they are in and easy to run locally or by Grinders - though not clear if dotty is resource issue.

**Comment by charliegracie** [comment]
2. I downloaded a JDK8 + OpenJ9 and I get a NullPointer exception everytime. This looks like a real issue on JDK8

**Comment by charliegracie** [comment]
1. dotty seems to work without any messages on JDK8 + OpenJ9 for me on my MAC no matter what heap setting I use.

**Comment by charliegracie** [comment]
This is the JDK8 + OpenJ9 build I used:
```
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (build 1.8.0_212-b03)
Eclipse OpenJ9 VM (build openj9-0.14.0, JRE 1.8.0 Mac OS X amd64-64-Bit Compressed References 20190417_217 (JIT enabled, AOT enabled)
OpenJ9   - bad1d4d06
OMR      - 4a4278e6
JCL      - 5590c4f818 based on jdk8u212-b03)
```

**Comment by charliegracie** [comment]
hmmm with the above JDK8 + OpenJ9 build I get a lot benchmarks throwing java.lang.ExceptionInInitializerError which appear to be caused by NullPointerExceptions

**Comment by charliegracie** [comment]
ok I moved up to the latest JDK8 + OpenJ9 release:
```
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (build 1.8.0_212-b04)
Eclipse OpenJ9 VM (build openj9-0.14.2, JRE 1.8.0 Mac OS X amd64-64-Bit Compressed References 20190521_245 (JIT enabled, AOT enabled)
OpenJ9   - 4b1df46fe
OMR      - b56045d2
JCL      - a8c217d402 based on jdk8u212-b04)
```

1. dotty seems to work with all settings I tried, including default
2. page-rank fails with the initial error reported with -Xmx512m and passes with -Xmx4g.  Looks like the machine is reported limited resources so the JVM picks a small heap size.
3. reactors fails with -Xmx512m and passes with -Xmx4g.  Looks like the machine is reported limited resources so the JVM picks a small heap size.

So to recap my findings there was definitely an issue with Scala and JDK8 + OpenJ9 in the 0.14 release but that appears to have been resolved in the 0.14.2 release. Also the machine running some of these benchmarks seem to not have enough resources.

**Comment by pshipton** [comment]
`-Xmx512m` is the default for Java 8. I suggest setting a bigger value to run the tests.

**Comment by pshipton** [comment]
Defaults https://www.eclipse.org/openj9/docs/openj9_defaults/
Also see https://github.com/eclipse/openj9/issues/3475

**Comment by NigelYiboYu** [comment]
I didn't run dotty.

`page-rank`  and `reactor` didn't run on zLinux in my experience. I'll try `-Xmx4g` later.

**Comment by smlambert** [comment]
Quick follow-up question, do openj9 defaults differ from hotspot / RI defaults?  

reactor run on test-packet-ubuntu1604-x64-1, jdk8 with openj9: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/1699

reactor run on same machine (test-packet-ubuntu1604-x64-1), jdk8 with hotspot : 
https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/1702

**Comment by pshipton** [comment]
> Quick follow-up question, do openj9 defaults differ from hotspot / RI defaults?

Yes, for jdk8 it is different. OpenJ9 jdk8 has the -Xmx512m hard limit for historical reasons, see #3475

**Comment by DanHeidinga** [comment]
> So to recap my findings there was definitely an issue with Scala and JDK8 + OpenJ9 in the 0.14 release but that appears to have been resolved in the 0.14.2 release. Also the machine running some of these benchmarks seem to not have enough resources.

@smlambert Is this issue still needed or are the machine configs / jvm options being handled under an Adopt issue?

**Comment by smlambert** [comment]
I will close this and refer to the Adopt issue, given that the other aspect is captured in #3475.
